network/p2p/scoring/subscription_validator_test.go

```diff
@@ -0,0 +1,267 @@
+func TestSubscriptionValidator_NoSubscribedTopic(t *testing.T) {
+	idProvider := mock.NewIdentityProvider(t)
+	sp := mockp2p.NewSubscriptionProvider(t)
+
+	sv := scoring.NewSubscriptionValidator(idProvider)
+	sv.RegisterSubscriptionProvider(sp)
+
+	// mocks peer 1 not subscribed to any topic.
+	peer1 := p2pfixtures.PeerIdFixture(t)
+	idProvider.On("ByPeerID", peer1).Return(unittest.IdentityFixture(), true)
+	sp.On("GetSubscribedTopics", peer1).Return([]string{})
+
+	// as peer 1 has not subscribed to any topic, the subscription validator should return no error.
+	require.NoError(t, sv.MustSubscribedToAllowedTopics(peer1))
+}
+
+// TestSubscriptionValidator_UnknownChannel tests that when a peer has subscribed to an unknown
+// topic, the subscription validator returns an error.
+func TestSubscriptionValidator_UnknownChannel(t *testing.T) {
+	idProvider := mock.NewIdentityProvider(t)
+	sp := mockp2p.NewSubscriptionProvider(t)
+
+	sv := scoring.NewSubscriptionValidator(idProvider)
+	sv.RegisterSubscriptionProvider(sp)
+
+	// mocks peer 1 not subscribed to an unknown topic.
+	peer1 := p2pfixtures.PeerIdFixture(t)
+	idProvider.On("ByPeerID", peer1).Return(unittest.IdentityFixture(), true)
+	sp.On("GetSubscribedTopics", peer1).Return([]string{"unknown-topic-1", "unknown-topic-2"})
+
+	// as peer 1 has subscribed to unknown topics, the subscription validator should return an error.
+	require.Error(t, sv.MustSubscribedToAllowedTopics(peer1))
+}
+
+// TestSubscriptionValidator_ValidSubscription tests that when a peer has subscribed to valid
+// topics based on its Flow protocol role, the subscription validator returns no error.
+func TestSubscriptionValidator_ValidSubscriptions(t *testing.T) {
+	idProvider := mock.NewIdentityProvider(t)
+	sp := mockp2p.NewSubscriptionProvider(t)
+
+	sv := scoring.NewSubscriptionValidator(idProvider)
+	sv.RegisterSubscriptionProvider(sp)
+
+	for _, role := range flow.Roles() {
+		peer := p2pfixtures.PeerIdFixture(t)
+		// allowed channels for the role excluding the test channels.
+		allowedChannels := channels.ChannelsByRole(role).ExcludePattern(regexp.MustCompile("^(test).*"))
+		sporkID := unittest.IdentifierFixture()
+
+		allowedTopics := make([]string, 0, len(allowedChannels))
+		for _, channel := range allowedChannels {
+			allowedTopics = append(allowedTopics, channels.TopicFromChannel(channel, sporkID).String())
+		}
+
+		// peer should pass the subscription validator as it has subscribed to any subset of its allowed topics.
+		for i := range allowedTopics {
+			idProvider.On("ByPeerID", peer).Return(unittest.IdentityFixture(unittest.WithRole(role)), true)
+			sp.On("GetSubscribedTopics", peer).Return(allowedTopics[:i+1])
+			require.NoError(t, sv.MustSubscribedToAllowedTopics(peer))
+		}
+	}
+}
+
+// TestSubscriptionValidator_SubscribeToAllTopics tests that regardless of its role when a peer has subscribed to all
+// topics, the subscription validator returns an error.
```

Comment:
this is because within the list of all topics, there are guaranteed to be some that are not allowed for the given role? can you make that clear in the comment.